Clocking API V2 - Rework

bradleyharden opened the pull request:
TODO: @bradleyharden, please update this list to reflect reality

#### Modules reworked, documented and ready for review (no risk of substantial changes):

- [ ] v2.rs
- [ ] v2/ahb.rs
- [ ] v2/apb.rs
- [ ] v2/dfll.rs
- [ ] v2/dpll.rs
- [ ] v2/gclk.rs
- [ ] v2/osculp32k.rs
- [ ] v2/pclk.rs
- [ ] v2/reset.rs
- [ ] v2/rtcosc.rs
- [ ] v2/types.rs
- [ ] v2/xosc.rs
- [ ] v2/xosc32k.rs
- [ ] ../../typelevel.rs
- [ ] ../../../../boards/feather_m4/examples/clocking_v2.rs

Updated on: 21.09.2022

bradleyharden committed:
dpll: Refactor predivider and remove enable Result return

bradleyharden committed:
gclk: Move output off value to Gclk from GclkOut

bradleyharden committed:
Add module-level docs and fix broken doc tests

bradleyharden committed:
Merge remote-tracking branch 'vccggorski/clocking-api-v2' into clock_v2_rebase

bradleyharden committed:
clock_v2: Fix merge problems

bradleyharden committed:
dpll: Fix prediv bug

bradleyharden committed:
Fix feather_m4 example

bradleyharden committed:
Fix errors introduced in merge commit

bradleyharden committed:
Add additional documentation

bradleyharden committed:
Further update the clock::v2 documentation

bradleyharden committed:
Flesh out gclk documentation

bradleyharden commented:
I noticed something while adding documentation, and I thought it was a discussion worth having here, rather than Matrix.

Right now, the `gclk::Tokens` struct is a collection of `GclkToken`s. We're not being consistent in our choice to include or exclude `Gclk` from type names. I think there are valid arguments for both approaches, so I wanted to see what you both think. 

In general, I think I lean more toward using the module name as a namespace and keeping the type names shorter. You can always add `gclk::` to disambiguate, if necessary. The only price you pay is two extra characters. And on top of that, I like the symmetry of having `gclk::Token`, `dpll::Token`, etc. Each module would have a `Token` type, a `Tokens` type, possibly a `SourceId` type, etc. I think it helps to highlight the fundamental architecture unifying all the different modules.

However, a potential counter argument is that multiple `Token` structs could be confusing if there is not an obvious context to tell you which is which. For example, I think the documentation for the `gclk` module would always refer to `gclk::Token` as merely `Token`, which could be confusing if you are linked to that documentation from somewhere else. It also could mean a little bit more typing for users when using multiple modules. But I also don't think users will need to reference anything other than `Gclk` itself very often, so maybe that's not a big concern.

For reference, here are the side by side names I would suggest, to compare them:

```text
GclkToken           gclk::Token
GclkTokens          gclk::Tokens
GclkId              gclk::Id
DynGclkId           gclk::DynId
GclkSourceId        gclk::SourceId
DynGclkSourceId     gclk::DynSourceId
GclkDivider         gclk::Divider
GclkDiv             gclk::Div8 (I'm not sure about these two. This is a first attempt)
Gclk1Div            gclk::Div16
```

Thoughts?

bradleyharden commented:
Pinging this again.

I also had a further realization about the naming issue, but I'm still conflicted.

If we remove `Gclk` from all the trait/type names, the `GclkId` types would become `gclk::Id0`, `gclk::Id1`, ... `gclk::Id11`. And what is now `DfllId` would become `dfll::Id`. In cases where there is just one clock, like the DFLL, `mod::Id` would be a type representing its identity. And in cases where there are multiple clocks, like the DPLLs or GCLKs, `mod::Id` would be a trait connecting a series of `Id0` through `IdN` types.

I like the uniformity of that structure, but my concern is again cases where context is lost. I'm not sure it would ever happen, but clearly `Gclk<Id4, Id>` is pretty confusing relative to `Gclk<Gclk4Id, DfllId>`.

There's also the issue of `Xosc1kId`, `Xosc32kId`, `OscUlp1kId` and `OscUlp32kId`. They would become `xosc32k::Id1k`, `xosc32k::Id32k`, `osculp32k::Id1k` and `osculp32k::Id32k`. I like the symmetry of `Id1k` and `Id32k` in both, but they become 5 characters longer to type, because of the module names.

Maybe we should just keep our current naming scheme and be redundant with the clock names. But at the very least, I think we should fix `gclk::Tokens`.

Thoughts?

bradleyharden commented:
I'm going to go ahead and change `gclk::Tokens` to `GclkTokens`.

I'm also going to change `GclkDiv` and `Gclk1Div` to `GclkDiv8` and `GclkDiv16`. I think that has a few advantages:
- They sort next to each other in the documentation. That's not the case for `GclkDiv` and `Gclk1Div`, which is a bit confusing.
- **Why** they are different is clearer (one is 8-bit, the other is 16-bit)
- If you only ever saw `GclkDiv` on its own, there's little to suggest that `Gclk1Div` even exists, which could cause confusion. But `GclkDiv8` sort of implicitly suggests that there might be other `GclkDiv` types.

bradleyharden committed:
gclk: Rename gclk::Tokens, GclkDiv and Gclk1Div

bradleyharden committed:
gclk: Fix typo

bradleyharden committed:
Fix feather_m4 example

bradleyharden committed:
gclk: Fix documentation typos

bradleyharden committed:
dpll: Improve naming consistency and add documentation

bradleyharden committed:
Improve language and explanations throughout

bradleyharden committed:
gclk: Correct missed changes in previous edits

bradleyharden committed:
pclk: Add documentation

vccggorski commented:
Namespace approach is sweet if someone is aware of it. Otherwise you end up with `IdX` types in the same namespace that conflict with one another. So that's the strong argument in favour of redundant, longer names.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,540 @@
+//!
+//! ### `Enabled` wrapper
+//!
+//! Bus clocks and leaves in the clock tree cannot be used as a source for
```
I think this _might_ be quite difficult to comprehend for a bystander. Maybe we should have an extra reference to clock system diagram from a datasheet here?

bradleyharden replied:
Yeah, it might be useful to snip relevant figures from the datasheet and put them in the docs.

glaeqen replied:
Maybe [some asciiflow?](https://asciiflow.com/) :D That might be a little over the top thing tho

bradleyharden replied:
What was this comment referring to again?

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,540 @@
+//! with an associated `Id` type of [`DfllId`]. While the [`Dfll<M>`] type
+//! itself needs the type parameter `M` to track its loop [`Mode`](dfll::Mode),
+//! downstream clocks don't need to know or care which mode the DFLL is using.
+//! The `Id` types serve to erase such configuration, representing only the
```
Erasure of irrelevant details hits nail on its head. This paragraph shows how difficult it is to properly elaborate on what's the relation and purpose of all these concepts/abstractions. Property (of erasing irrelevant details) in order not to poison downstream types with redundant information (modes but also user count of dependences) is mentioned but maybe it should be stressed even more + showing an example and counter-example of what would happen if we _just_ allowed types to nest (`Enabled<X<Enabled<.., 8>, 3>`).

What is missing, I think, is a chapter explaining `*SourceId` trait family which ties everything together and limits which `Source` can be used against which clock *via* `*Id` type implementing it (+ providing runtime relevant constants (`Dyn*SourceId` enums) that are supposed to be fed into HW mapped registers for setting clock sources and so on).

bradleyharden replied:
Could you look at the existing docs for Source::Id? I plan to elevate that to the top level docs

glaeqen replied:
I guess we effectively went through this point over Matrix

bradleyharden replied:
@vccggorski, please see the latest documentation and see if you are satisfied. If not, please unresolve this.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,540 @@
+//!
+//! ### Bus clocks
+//!
+//! Bus clocks are fundamentally different from the other clock types. They are
```
I wouldn't mind sneaking in some terminology established in ATMEL's datasheets. I think they prefer to talk about asynchronous clocking domain and synchronous clocking domain (GCLK0 as MCLK -> CPU buses). I'm curious what you think about it but maybe sticking to the semantics of datasheet should ease a little a mental burden on a person that just tries to understand the clocking system. Or MCU architecture in general. Maybe it's just me and me very lackluster electronics background though.

bradleyharden replied:
Yeah, you're right. I just haven't read the docs in ages. Let me look through them and see if we can be more consistent

bradleyharden replied:
I think this is irrelevant now. Synchronous vs. asynchronous is not really important for the `clock` module. What makes bus clocks different is the fact that they share registers. I added a section on memory safety, which I think clarifies why the bus types exist. Please unresolve this if you think there's something left unaddressed.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,540 @@
+//! running at power-on reset, namely:
+//! - All of the [`AhbClks`]
+//! - Some of the [`ApbClks`]
+//! - The 48 MHz [`Dfll`], running in [`OpenLoop`] mode
```
`Dfll` link is broken; I'd suggest going through the warnings and rendered docs to look up all the faulty misgenerated hyperlinks.

bradleyharden replied:
I haven't bothered with that, because so many of the unfinished modules are loaded with broken links. Is there a way to filter the warnings by module?

glaeqen replied:
I don't think so.. But looking at the output, 36 out of 39 doc warnings come from clock module anyway.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,540 @@
+//! - The [`OscUlpBase`] clock, which is always running and cannot be disabled.
+//!   It can feed the derived [`OscUlp1k`] and [`OscUlp32k`] clocks.
+//!
+//! Note that bus clocks are not wrapped in [`Enabled`], because they can never be
```
The way I always made a mental distinction is by referring to clock domains.

![clocking domains](https://user-images.githubusercontent.com/80407822/179518552-8e6b0c4e-77a9-461f-b110-223869987997.png)

Bus clocks power CPU buses so CPU can even interact with the peripheral. This is why it's synchronous because it follows MCLK/GCLK0. Asynchronous clocking domain powers peripherals as completely seperate pieces of HW. Bus clocks are like `PChs`; they are tokens of correctness for abstractions representing the HW they power - both from synchronous and asynchronous domains. They have 1:1 relations to them; thus whole idea of `Enabled` does not apply. Obviously you know that but maybe this can be conveyed to the user to harden his understanding of the system.

bradleyharden replied:
The 1:1 relationship is a really good point. I stressed that a bit in the `Pclk` documentation. I think that's a clearer way to explain it than what I have now. I'll restructure it. `Enabled` is only needed when you have a 1:N relationship.

bradleyharden replied:
Again, I think sync vs. async clocks are irrelevant. What matters is 1:1 vs. 1:N clocks, which is now heavily emphasized in the documentation. Please unresolve if there is something more.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,540 @@
+//! disable it. However, the same is not true for [`Gclk0`]. Because the master
+//! clock has no representation in this module, the [`Counter`] for
+//! [`EnabledGclk0`] can *never* be reduced to [`U0`], which neatly models that
+//! the master clock can never be disabled.
```
This paragraph kinda suggests then that you cannot change the source for a GCLK0? Hmm

bradleyharden replied:
Maybe I can add a comment referring to the actions you *can* take with `EnabledGclk0`.

glaeqen replied:
I reread this paragraph few times and what was confusing for me in the phrasing of it was that I kept reading it as "Gclk0 can be configured to use a different base clock, (...); however, the same is not true for Gclk0"). Which obviously is illogical. What you mean is that "(...) allow you to reduce the Dfll's counter from U1 to U0 (..); however the same is not true for Gclk0". Probably it's just my brain playing tricks on me :D

bradleyharden replied:
I think this has been addressed now.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,540 @@
+//!     └── GCLK0 (100 MHz)
+//!         ├── Master clock (100 MHz)
+//!         ├── SERCOM0 peripheral clock
+//!         └── Ouput to GPIO pin
```
Typo

bradleyharden replied:
Fixed

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,540 @@
+//! └── GCLK0 (48 MHz)
+//!     └── Master clock (48 MHz)
+//!
+//! OSCULP base clock
```
This dangling `OSCULP` can be a little confusing; looks like an extra description, not like a separate 1-leaf branch of a clock tree. Not sure.

bradleyharden replied:
Yeah, I wondered that myself. I think we can just omit it. I don't think that would confuse people.

bradleyharden replied:
Removed throughout.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
````diff
@@ -1,98 +1,540 @@
+//! First, let's import some of the necessary types. We will see what each type
+//! represents in turn.
+//!
+//! ```
````
All examples should be marked with `no_run`; otherwise `cargo doc --test` will fail at linking trying to run this code. Apart from that it produces this warning marker in a generated documentation saying that the example is not tested.

glaeqen replied:
Consider using `#` in docs code examples to hide some lines from being rendered but which are essential for the code piece to compile.

bradleyharden replied:
I was going to do that, but I was worried the docs would get huge within the code, all for a bunch of repeat copies of the same code commented out. But you make a good argument about stale documentation and doc tests.

glaeqen replied:
Not sure what can be done about it. It'd be nice to have like a "common hidden" code shared between code blocks. Like, let's say, imports and clock struct initialization and then you can have short example code blocks reusing it. Maybe I missed something like that in Cargo book.

bradleyharden replied:
I've made all the examples in the main documentation `no_run` instead of `ignore`. I'll leave this open until I update the `gclk`, `dpll` and `pclk` modules.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,540 @@
+//! aim to create an instance of `Dpll0<Xosc0Id>`.
+//!
+//! Only certain clocks can drive the DPLL, so `I` is constrained by the
+//! [`DpllSourceId`] trait. Specifically, only the [`Xosc0Id`], [`Xosc1Id`],
```
Ok, you mention the existence of `*SourceId` traits here. Maybe you didn't want to increase the complexity of the module docs in the chapter explaining concepts but I think it is quite important to mention it there maybe.

bradleyharden replied:
I will elevate discussion of Source::Id. That's a good call.

bradleyharden replied:
I think this has been addressed now.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
````diff
@@ -1,98 +1,540 @@
+//! alias for `Enabled<Xosc0<M>, N>`. In this case, we get
+//! `EnabledXosc0<CrystalMode, U0>`.
+//!
+//! ```ignore
````
This code should be compilable as part of the `cargo doc --test`, so I'm in favour of using `no_run` here. This might require some amount of boilerplate hidden away with `#` to make it work but I think it is worth in order to prevent examples from rotting.

bradleyharden replied:
Yeah. I was worried about bloat and was being lazy. I can add that

bradleyharden replied:
Closing this as a duplicate of the comment above.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,540 @@
+//! Thus, to create an instance of `Dpll0<XoscId0>`, we must provide the
+//! [`EnabledXosc0`], so that its [`U0`] type parameter can be incremented to
+//! [`U1`]. The `Dpll::from_xosc0` method takes ownership of the
+//! [`EnabledXosc0`] and returns it with this modified type parameter.
```
This is the most powerful part of the API that, I'd argue, should be stressed maybe even more. So, "Thanks to the user counter being incremented from U0 to U1 for Xosc0 from now on, unless `Dpll` is disabled, `EnabledXosc0` is protected from being disabled".

bradleyharden replied:
I missed this before. I will add a sentence or two to emphasize this more.

bradleyharden replied:
Done

glaeqen commented on hal/src/thumbv7em/clock/v2/dpll.rs:
```diff
@@ -630,16 +919,16 @@ where
+        self.0.token.is_ready()
     }
 }
 
```
In `Source` implementation for `EnabledDpll`, is a `GclkSourceId` restriction on Id needed?

bradleyharden replied:
I have no idea where that came from. I've removed it. I'll push the update soon.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
````diff
@@ -1,98 +1,540 @@
+//! let (gclk_out0, gclk0) = GclkOut::enable(tokens.gclk_io.gclk_out0, pins.pb14, gclk0);
+//! ```
+//!
+//! We have arrived at our final, desired clock tree. Putting the whole example
````
Really cool piece of writing, great job :partying_face:

glaeqen commented on hal/src/thumbv7em/clock/v2/gclk.rs:
```diff
@@ -545,37 +729,82 @@ impl GclkSourceId for Xosc32kId {
+/// prevents it from ever being disabled. Accordingly, we also provide a few
+/// special methods on [`EnabledGclk0`] to configure the `Gclk` while it is
+/// actively running.
+pub type EnabledGclk0<I, N = U0> = EnabledGclk<Gclk0Id, I, N>;
```
Maybe Gclk0 should have a default amount of users == `U1` ?

bradleyharden replied:
Good call. That would be more useful. We should point it out in the docs too.

bradleyharden replied:
Fixed

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -178,12 +638,11 @@ impl<T, N: PrivateDecrement> PrivateDecrement for Enabled<T, N> {
-/// Test
-pub fn test() {
+#[allow(dead_code)]
+fn test() {
```
I guess it'd be nice to get rid of this function in the long run in favour of comprehensive doc-tests suite.

bradleyharden replied:
Yes, that was the plan. But it's nice to have in the code for now, because Ruat Analyzer catches errors in real code immediately. Maybe there's a way to continuously check doc code as well?

glaeqen replied:
Doc code can be checked with `no_run` as was said in other comments, boilerplate is a difficult issue to solve unfortunately.

bradleyharden replied:
Closing this. We will delete the `test` function before release. I'm keeping it for now, because rust-analyzer doesn't point out failing doc examples.

glaeqen commented on hal/src/thumbv7em/clock/v2/dpll.rs:
```diff
@@ -1,29 +1,167 @@
+//! ## Overview
 //!
-//! There are two Dplls that are available
+//! The `dpll` module prodives access to the two digital phase-locked loops
```
typo

bradleyharden replied:
Fixed

glaeqen commented on hal/src/thumbv7em/clock/v2/dpll.rs:
```diff
@@ -489,134 +777,135 @@ where
+    ///
+    /// [`new`]: Dpll::new
     #[inline]
-    pub unsafe fn force_enable(mut self) -> EnabledDpll<D, I> {
```
I created this function in the first place because I need Dpll being creatable in "slightly out-of-spec" configuration outputting 90 MHz. So I need a way to ignore this check :smile:

bradleyharden replied:
Ah, ok. Maybe we should have an `enable_unchecked` version?

glaeqen replied:
That'd nice, I agree

bradleyharden replied:
I've added this. Pushing soon.

glaeqen commented on hal/src/thumbv7em/clock/v2/gclk.rs:
```diff
@@ -545,37 +729,82 @@ impl GclkSourceId for Xosc32kId {
+/// actively running.
+pub type Gclk0<I> = Gclk<Gclk0Id, I>;
+
+/// Type aliase for the corresponding [`EnabledGclk`]
```
typo

bradleyharden replied:
I thought I fixed that one? Were you reviewing the latest version?

glaeqen replied:
Yeah, I double checked it now. Commit `c4d893cc7`, `gclk.rs:792`, `aliase` is still there.

bradleyharden replied:
Fixed

glaeqen commented on hal/src/thumbv7em/clock/v2/pclk.rs:
```diff
@@ -105,10 +178,6 @@ pub mod ids {
 
     pub use super::super::dfll::DfllId;
     pub use super::super::dpll::{Dpll0Id, Dpll1Id};
-    pub use super::super::gclk::{
```
I really like how clever, clean and future-proof is the re-export approach.

glaeqen reviewed (changes requested):
In retrospect I think a lot of my comments are little over the top and probably can be taken with a grain of salt. Looks really good although sometimes amount of links in rendered document looks hilarious :P I think my biggest complaints would be:

- Lack of explanation of Dyn type semantics or at least a link to the existing explanation in other part of the codebase
- SourceId is quite important considering it sits in between Id types and Source types; latter does not implement any usage restrictions so it's a essential that a "user" of a type (usually clock type constructors in that case) put extra limitation on `Source::Id`
- It'd be nice to get doc-tests with `no_run` going (and part of the CI in the long run)
- typos and broken links - probably quite mundane but needs to be done. I could probably help with you with that but I don't want to get in the way in case you're currently developing something.

I guess you're currently working on modules like AHB, XOSC and so on (their docs got sometimes completely outdated) or do you need a hand in that?

nilfit commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,540 @@
+//!     - Peripheral channel clocks ([`pclk`])
+//!     - The real-time clock oscillator ([`rtcosc`])
+//!     - External clock ouputs ([`GclkOut`])
+//! - Bus clocks derived from the processor's master clock:
```
master clock -> main clock to be consistent with the datasheet MCLK. The only master clocks I find there are in I2C, I2S and ethernet peripherals.

bradleyharden replied:
Done

bradleyharden committed:
Update main documentation based on feedback

bradleyharden committed:
dpll: Add enable_unchecked and remove unnecessary trait bound

bradleyharden committed:
dpll: Fix typo

bradleyharden committed:
gclk: Fixe typo and change EnabledGclk0 default N = U1

bradleyharden committed:
Fix typo and improve documentation

bradleyharden committed:
gclk: Refactor handling of GCLK_IO and update documentation

`GclkIn` was a 1:1 clock, but it used `Enabled` and `Source` as if it
where a 1:N clock. Instead, remove `GclkIn` completely and modify `Gclk`
to use `from_source` or `from_pin` constructors.

Allow multiple `GclkOut`s per `Gclk`, since this is allowed by the
hardware. Doing so removes the need for any `GclkOutToken`s.

Consolidate the `gclkio` module with the `gclk` module, because it has
been reduced to only a single trait and type.

Update the documentation and code throughout to match the changes.

bradleyharden committed:
Refactor `dpll` to match `gclk` and update docs throughout

bradleyharden committed:
xosc: Update module to improve naming and handle more cases

bradleyharden committed:
xosc: Finalize `xosc` module and update docs & examples

bradleyharden committed:
dpll: Fix minor documentation issues

bradleyharden committed:
xosc: Fix minor issues

bradleyharden committed:
Update 32 kHz oscillator modules and documentation

bradleyharden committed:
Update feather_m4

bradleyharden committed:
Fix minor documentation issues

bradleyharden committed:
Update AHB and APB module documentation

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,859 @@
+//! being modified or disabled once it has been connected to a consumer. We
+//! don't need to permanently hold the `Producer` object in place to do that.
+//!
+//! It is possible to build a `clock` API based on the borrow checker, but it
```
Is it? You'd be able to mutate/disable parent clock as soon as you never touch the object representing the child clock (which borrows the parent enabled clock, because last usage == end of borrow). If object goes out of scope, it does not mean that it disables a clock that the object represents (it would be undesirable anyway). Borrowing is very weak mechanism to represent this idea.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,859 @@
-//! an ATSAMD's clocking system in a safe and expressive manner.
+//! Only certain clocks can drive the DPLL, so `I` is constrained by the
+//! [`DpllSourceId`] trait. Specifically, only the [`Xosc0Id`], [`Xosc1Id`],
+//! [`Xosc32kId`] and [`GclkId`] types implement this trait.
```
`GclkId` is not really a type, but I guess that is sort of nitpicking and more precision would muddy the waters in a sentence. So I assume it is fine the way it is

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,859 @@
+//! [`DpllToken`] when creating an instance of `Dpll`. However, unlike before,
+//! we are creating a new clock-tree relationship that must be tracked by the
+//! type system. Because DPLL0 will now consume XOSC0, we must [`Increment`]
+//! the [`Enabled`] [`Counter`] for [`EnabledXosc0`].
```
`Counter` is `Enabled`? Hmm :/

glaeqen replied:
I'm not sure about `we must` phrasing used together with `Increment` and `Decrement` later on. Does it suggest that user should decrement/increment some counters themselves or is it obvious that `from_source` and `free_source` do that themselves?

glaeqen replied:
> Because DPLL0 will now consume XOSC0, the `Counter` in `EnabledXosc0` is going to get incremented.

Now when I think about it, `consume` can be misleading for some ppl thinking about Rust's `consuming` of a variable? Hmm. I might be overreacting though.

glaeqen commented on hal/src/thumbv7em/clock/v2.rs:
```diff
@@ -1,98 +1,859 @@
+//! been incremeneted to `U1`, the `EnabledXosc0` can no longer be modified or
+//! disabled. All further code can guarantee this invariant is upheld. To modify
+//! the `EnabledXosc0`, we would first have to use `Dpll::free_source` to
+//! disable the DPLL and [`Decrement`] the `Counter` back to `U0`.
```
> (..) and `Counter` would be `Decrement`ed back to `U0`. 

I'm referring to previous comment of mine.

Although we also discussed removal of `Counter` trait so probably that will get totally reworded

glaeqen commented on hal/src/thumbv7em/clock/v2/gclk.rs:
```diff
@@ -723,54 +1198,150 @@ where
-        let mut config = self.0.div(div);
-        config.token.set_div(div);
-        Enabled::new(config)
+    pub fn freq(&self) -> Hertz {
```
Redundant, All `EnabledGclk`s implement `Source` trait that have this function.

vccggorski commented:
Merged manually with restored linear history
